Remove codebehind files

Tragedian opened the pull request:
Removes the generated codebehind files from the project structure. Generated files are now written to a subdirectory in the `/obj/` folder.

- Change MSBuild targets for build and clean to generate files as a build artefact prior to compilation.
- Add MSBuild spec project to verify generated results across scenarios
- Add assembly analyzer project which determines what scenarios have been compiled into an assembly 

There will need to be some advice given on upgrading, i.e. all feature.cs files should be removed from the project folders, and the documentation at https://specflow.org/documentation/Generate-Tests-from-MsBuild/ could possibly benefit from some information about the few properties available to customize the process.

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected).

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I've added tests for my code.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added an entry to the changelog

Tragedian commented:
I can see a CI build error, but I'm not able to reproduce it locally. Is there something special about `build.ps1` when it runs as part of an automated build?

SabotageAndi commented:
Just FYI, I am still on vacation until next week. 
So it will take a while until I can have a look at the PR.

About CI: it is doing a clean build. Try a `git clean -fdx .` and make then a build. Looks like it is a build order problem.

Tragedian commented:
Cleaning with git reproduced the problem for me. Massive thanks. Not to get to the bottom of the error.

SabotageAndi commented on SpecFlow.Tools.MsBuild.Generation/build/SpecFlow.Tools.MsBuild.Generation.props:
```diff
@@ -1,82 +1,27 @@
-    <!-- Support for Visual Studio Incremental Build
-        https://github.com/techtalk/SpecFlow/issues/1319
-     -->
-    <UpToDateCheckInput Include="@(SpecFlowFeatureFiles)" />
```
This 2 itemgroups are important that VS detects file changes and starts a build for the project if they are changed.

Tragedian replied:
Restored the `UpToDateCheckInput`; I had originally planned to put it back once I got the build-time generation to work, but completely forgot I had removed it in the first place.

Tragedian replied:
The other item groups shouldn't be required as they're no longer project items. Do we have any manual test cases to check the VS integration?

SabotageAndi commented on SpecFlow.Tools.MsBuild.Generation/build/SpecFlow.Tools.MsBuild.Generation.props:
```diff
@@ -1,82 +1,27 @@
+    <SpecFlowTasksAssembly Condition="'$(SpecFlowTasksAssembly)' == ''">..\tasks\net471\SpecFlow.Tools.MsBuild.Generation.dll</SpecFlowTasksAssembly>
   </PropertyGroup>
 
-  <Import Project="$(SpecFlow_CpsExtensionDesignTimeTargetsPath)" Condition="'$(DesignTimeBuild)' == 'true' " />
```
This is important for VS integration of the feature files and that SingleFileGenerators aren't added.

Tragedian replied:
I wasn't sure this was still needed with the generated files no longer being included in the project. I'm not familiar with the extension points for the VS Project system, so a little guidance here would be very appreciated.

SabotageAndi commented on SpecFlow.Tools.MsBuild.Generation/build/SpecFlow.Tools.MsBuild.Generation.props:
```diff
@@ -1,82 +1,27 @@
-    <CustomAdditionalCompileInputs Include="@(SpecFlowFeatureFiles->'%(CodeBehindFile)')" />
-
+    <SpecFlowFeature Include="**/*.feature" />
+    <SpecFlowFeature Include="**/*.feature.xls" />
```
The ending was `xslx` and not `xls`

Tragedian replied:
Oof. My bad; I should be able to fix that right up.

SabotageAndi commented on SpecFlow.Tools.MsBuild.Generation/build/SpecFlow.Tools.MsBuild.Generation.targets:
```diff
@@ -1,145 +1,76 @@
+    
+    <ItemGroup>
+      <SpecFlowGeneratedCodeFile Include="$(SpecFlowGeneratedCodePath)**/*.feature$(DefaultLanguageSourceExtension)" />
+      <SpecFlowGeneratedCodeFile Include="$(SpecFlowGeneratedCodePath)**/*.feature.xls$(DefaultLanguageSourceExtension)" />
```
Wrong ending for Excel.

Tragedian replied:
Should be fixed now.

Tragedian commented:
@SabotageAndi 

> I had a look now at the PR. As I understand it, it will generate the feature.cs files now in the `obj` folder.

Glad you could gather that just from the code. It is as you surmise: the generated files are placed in an directory in the intermediary output folder.

> * Why did you delete the `TestFiles`? They are part of the `Parser smoke tests`

I think I did this by accident when removing the `feature.cs` files. Thanks for clarifying their purpose; I have reverted their deletion.

> * You split up the hooks file. Any concrete reason for it? It was on purpose that there is only one class with all the hooks. Maintaining hook order over a lot of files is a pain.

I split the hooks by applying the single responsibility principal: there should only be one reason to change a class and all the hook functions do quite different things. I'm not going to die on that hill though, so if you'd prefer them merged into a singular hooks class I can make it so.

> * Does jumping from the test explorer to the feature file still work?

**Yes!** This was something I tested repeatedly.

> * Does debugging with breakpoints in the feature file still work?

I honestly had no idea you could do that. Breakpoints added to a feature file don't highlight the line like they do in a regular code file and I can't break into them when debugging the test:

![image](https://user-images.githubusercontent.com/3265994/63369769-908a8600-c378-11e9-8dcc-3d45408d84bc.png)

Tragedian commented:
@SabotageAndi I did an update from the upstream but it looks like there's something going on with the CI build retrieving the cucumber submodule. Not sure if there's anything I can do from my end?

SabotageAndi commented:
Sorry, that's on my. I forced pushed the branch we are using in the submodule. I am working on it to fix it in master.

Tragedian commented:
@SabotageAndi The branch builds now, but I'm still seeing really weird behaviour from the CI trying to start the tests for each platform. I hope you can help.

SabotageAndi commented:
We are trying to fix the release pipelines on Linux currently. Could be that we broke something on in the Windows release also.
I will check it on Monday.

Tragedian commented:
@SabotageAndi Looks like the build is working again! I've got some Spec failures to investigate and fix.

Tragedian commented:
I've fixed a number of simple issues and the last thing in the way of a clean set of unit tests seems to be a problem relating to the `XUnitAssemblyFixture` not being available as an argument to a test fixture. It works fine in all the test scenarios, except those under `netcoreapp21`;  the `netcoreapp30` tests and those targeting the .NET Framework seem to function normally.

It's probably going to be a really simple fix to a build property I've been working with, but I have to find it first.

SabotageAndi commented:
I would check your changes in https://github.com/techtalk/SpecFlow/pull/1691/files#diff-3b9a1f9868e438e5cbb6ef6048743e20. If the class is missing, the file is not added to the compile- itemgroup.

Tragedian commented:
Sorry I've been quiet on this work for a while. I've been beset by illness and bereavement, but now I'm back to fix this thing up.

@SabotageAndi  The problem I'm currently facing is one where the I get the error:

> The following constructor parameters did not have matching fixture data: XUnitAssemblyFixture assemblyFixture

Having decompiled the assemblies built by the test, I can see the class was included in the built assembly. The weirdest thing is this only happens for the "XUnit v2 unit test provider" under the `Netcoreapp21_New_CSharp` configuration. All other configurations work!

Any thoughts as to what could be going on? I've got to update from master yet, which may change things, but it would be good if you have any thoughts to what is special about the .NET Core 2.1 flavour.

SabotageAndi commented on Plugins/TechTalk.SpecFlow.xUnit.Generator.SpecFlowPlugin/build/SpecFlow.xUnit.props:
```diff
@@ -1,12 +1,22 @@
+    <SpecFlowXUnitGeneratorPluginAssembly Condition="'$(SpecFlowXUnitGeneratorPluginAssembly)' == ''">$(_SpecFlowXUnitGeneratorPluginPath)TechTalk.SpecFlow.xUnit.Generator.SpecFlowPlugin.dll</SpecFlowXUnitGeneratorPluginAssembly>
+
+    <_SpecFlowXUnitRuntimePluginPathBase>$(MSBuildThisFileDirectory)../lib/</_SpecFlowXUnitRuntimePluginPathBase>
+    <_SpecFlowXUnitRuntimePluginPath Condition=" '$(MSBuildRuntimeType)' == 'Core'" >$(_SpecFlowXUnitRuntimePluginPathBase)netstandard2.0/</_SpecFlowXUnitRuntimePluginPath>
```
for the runtime plugin you have to check for the targetframework and not for the MSBuild version. You can compile a .NET Framework project with `dotnet build`. This would be then a `Core` MSBuildRuntimeType for a .NET Framework project

Tragedian replied:
This _might_ be the cause of my problems. I still need to resolve a change in how the XUnit assembly-hooks file is added.

SabotageAndi commented on SpecFlow.Tools.MsBuild.Generation/build/SpecFlow.Tools.MsBuild.Generation.props:
```diff
@@ -1,77 +1,32 @@
-    <SpecFlowObsoleteCodeBehindFiles Include="**\*.feature$(DefaultLanguageSourceExtension)" Exclude="@(SpecFlowFeatureFiles->'%(CodeBehindFile)')" />
-
+    <SpecFlowFeature Include="**/*.feature" />
+    <SpecFlowFeature Include="**/*.feature.xlsx" />
```
in the meantime we removed the includition of the `feature.xlsx` files.

Tragedian replied:
Easily removed. I didn't realise the Excel support had been dropped. I can't say I had ever used it though.

SabotageAndi commented on SpecFlow.Tools.MsBuild.Generation/build/SpecFlow.Tools.MsBuild.Generation.targets:
```diff
@@ -1,167 +1,69 @@
+  -->
+  <Target
+    Name="GenerateSpecFlowCodeFiles"
+    BeforeTargets="BeforeCompile"
```
we had problems with the `BeforeTargets` and `AfterTargets`. That's why we are using the `PropertyGroup`.
Also it makes it easier to have the right order. We have multiple points where we adjust this order.

SabotageAndi commented on Tests/TechTalk.SpecFlow.Specs/TechTalk.SpecFlow.Specs.csproj:
```diff
@@ -63,35 +84,38 @@
-  <Target Name="PreBuild">
-    <MSBuild Projects="..\..\SpecFlow.Tools.MsBuild.Generation\SpecFlow.Tools.MsBuild.Generation.csproj" ContinueOnError="false" Properties="TargetFramework=$(SpecFlow_Core_Tools_TFM)"/>
-    <MSBuild Projects="..\TechTalk.SpecFlow.Specs.Generator.SpecFlowPlugin\TechTalk.SpecFlow.Specs.Generator.SpecFlowPlugin.csproj" ContinueOnError="false" Properties="TargetFramework=$(SpecFlow_Core_Tools_TFM)"/>
+  <Target Name="PreBuild" BeforeTargets="Build">
```
That's one of the target where we had problems with the `BeforeTargets`.

Tragedian commented:
Master has moved on a lot since I started this work. I think at this stage it's going to be easier to take a fresh branch and re-apply my changes.
